pipeline_components/aiplatform/aiplatform/custom_train_job/component.py

```diff
@@ -0,0 +1,167 @@
+        valid_data (Dataset): Validation data (passed as an argument to task script)
+        test_data (Dataset): Test data (passed as an argument to task script).
+        staging_bucket (str): Staging bucket for CustomTrainingJob.
+        project_location (str): location of the Google Cloud project.
+        project_id (str): project id of the Google Cloud project.
+        model_display_name (str): Name of the new trained model version.
+        train_container_uri (str): Container URI for running task script.
+        serving_container_uri (str): Container URI for deploying the output model.
+        model (Model): Trained model output.
+        metrics (Metrics): Output metrics of trained model.
+        requirements (List[str]): Additional python dependencies for training script.
+        job_name (str): Name of training job.
+        hparams (Dict[str, str]): Hyperparameters (passed as a JSON serialised argument
+            to task script)
+        replica_count (int): Number of replicas (increase for distributed tasks).
+        machine_type (string): Machine type of compute.
+        accelerator_type (string): Accelerator type (change for GPU support).
+        accelerator_count (string): Accelerator count (increase for GPU cores).
+
+    Returns:
+        parent_model (str): Resource URI of the parent model (empty string if the
+            trained model is the first model version of its kind).
+    """
+    import json
+    import logging
+    import os.path
+    import time
+    import google.cloud.aiplatform as aip
+
+    logging.info(f"Checking if parent model with name {model_display_name} exists")
+    models = aip.Model.list(filter=f"displayName={model_display_name}")
+
+    # check if a model with the same name already exists
+    if len(models) == 0:
+        logging.info("No parent model found.")
+        parent_model = None
+        is_default_version = True
+    elif len(models) == 1:
+        parent_model = models[0].resource_name
+        is_default_version = False
+        logging.info(f"Parent model found: {parent_model}")
+    else:
+        raise RuntimeError(
+            f"Multiple models with name {model_display_name} were found."
+        )
+
+    logging.info(f"Using task: {task_uri}")
+    script_path = "/gcs/" + task_uri[5:]
+    if not os.path.exists(script_path):
+        logging.error("Task script was not found!")
+        raise ValueError(
+            f"Task script was not found. Check if the path is correct: {task_uri}"
+        )
+
+    job = aip.CustomTrainingJob(
+        project=project_id,
+        location=project_location,
+        staging_bucket=staging_bucket,
+        display_name=job_name if job_name else f"Custom job {int(time.time())}",
+        script_path=script_path,
+        container_uri=train_container_uri,
+        requirements=requirements,
+        model_serving_container_image_uri=serving_container_uri,
+    )
+    cmd_args = [
```

Comment:
I think we should be less prescriptive with the command-line arguments. Ideally these would be provided as a list by the user as input to the component (but I see that is tricky given that it contains artifact paths). Could we output the paths as strings from previous components, construct the args in the pipeline definition and then pass into this components as `args: List[str]` ?

Reply:
unfortunately can't use op outputs (or pipeline params) in nested objects, leads to compilation error.

Reply:
tested:
```python
train_dataset_uri = (
    extract_bq_to_dataset(
        #...
    )
    .after(data_cleaning)
    .set_display_name("Extract train data to storage")
).outputs['dataset_gcs_uri']

train_args = [
    "--train_data", train_dataset_uri,
    # ...
]

train_model = custom_train_job(
    train_script_uri=train_script_uri,
    args=train_args,
    # ...
).set_display_name("Train model")
```

Error: `TypeError: Object of type PipelineParam is not JSON serializable`